### PR TITLE
Add NetworkPolicy to Nginx 

### PR DESCRIPTION
**Pull request by griffithscg**

### CODE REVIEW

File: stable/nginx/nginx/templates/NetworkPolicy.yaml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+apiVersion: networking.k8s.io/v1
+kind: NetworkPolicy
+metadata:
+  name: policy-demo 
+  namespace: default
```
Comment (Mansalu) [inline stable/nginx/nginx/templates/NetworkPolicy.yaml:5]:
```suggestion
```

I think we don't want this the policy needs to exist in the correct SLATE group namespace.

SLATE API should just handle this during the application install process? Perhaps, @cnweaver you can confirm?

File: stable/nginx/nginx/templates/NetworkPolicy.yaml [hunk 2]
```diff
@@ -0,0 +1,18 @@
+apiVersion: networking.k8s.io/v1
+kind: NetworkPolicy
+metadata:
+  name: nginx-{{ .Values.Instance }}-network-policy
```
Comment (Mansalu) [inline stable/nginx/nginx/templates/NetworkPolicy.yaml:4]:
```suggestion
  name: {{ template "nginx-fullname" . }}
```

Let's change this to conform to the object naming convention.

File: stable/nginx/nginx/templates/NetworkPolicy.yaml [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    instance: {{ .Values.Instance | quote }}
+spec:
+  podSelector: {}
+  policyTypes:
+  - Ingress
+  - Egress
+  ingress:
+  - from:
+    - ipBlock:
+        cidr: 155.101.8.34/32
```
Comment (Mansalu) [inline stable/nginx/nginx/templates/NetworkPolicy.yaml:23]:
```suggestion
        cidr: {{ .Values.NetworkPolicy.AllowedCIDR }}
```

Let's template this.

File: stable/nginx/nginx/templates/NetworkPolicy.yaml [hunk 4]
```diff
@@ -0,0 +1,23 @@
+apiVersion: networking.k8s.io/v1
```
Comment (Mansalu) [inline stable/nginx/nginx/templates/NetworkPolicy.yaml:1]: {{ if .Values.NetworkPolicy.Enabled }}
```suggestion
apiVersion: networking.k8s.io/v1
```

File: stable/nginx/nginx/templates/NetworkPolicy.yaml [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+      chart: {{ template "nginx.chart" . }}
+      release: {{ .Release.Name }}
+      instance: {{ .Values.Instance | quote }}
+  policyTypes:
+  - Ingress
+  - Egress
+  ingress:
+  - from:
+    - ipBlock:
+        cidr: 155.101.8.34/32
```
Comment (Mansalu) [inline stable/nginx/nginx/templates/NetworkPolicy.yaml]:
```suggestion
        cidr: 155.101.8.34/32 
{{ end }}
```

File: stable/nginx/nginx/values.yaml [hunk 6]
```diff
@@ -22,9 +22,14 @@ Ingress:
   # The class of the ingress controller to use. 
   # For SLATE this should be 'slate'. 
   Class: slate
-  
+
+#Defines what IP ranges can access the app
+#If disabled all IP addresses can access the app
+NetworkPolicy:
+  Enabled: false
+  AllowedCIDR: 10.0.0.0/255
```
Comment (Mansalu) [inline stable/nginx/nginx/values.yaml:30]:
```suggestion
  AllowedCIDR: 10.0.0.0/8
```

This is not a valid CIDR.

File: stable/nginx/nginx/templates/NetworkPolicy.yaml [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+  podSelector:
+    matchLabels:
+      app: {{ template "nginx.name" . }}
+      chart: {{ template "nginx.chart" . }}
+      release: {{ .Release.Name }}
+      instance: {{ .Values.Instance | quote }}
+  policyTypes:
+  - Ingress
+  - Egress
+  ingress:
```
Comment (Mansalu) [inline stable/nginx/nginx/templates/NetworkPolicy.yaml:21]:
```suggestion
  ingress:
```

Comment (Mansalu) [inline stable/nginx/nginx/templates/NetworkPolicy.yaml:21]:
```suggestion
  egress:
  - to:
    - ipBlock:
        cidr: {{ .Values.NetworkPolicy.AllowedCIDR }}
  ingress:
```

File: stable/nginx/nginx/templates/NetworkPolicy.yaml [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+  podSelector:
+    matchLabels:
+      app: {{ template "nginx.name" . }}
+      chart: {{ template "nginx.chart" . }}
+      release: {{ .Release.Name }}
+      instance: {{ .Values.Instance | quote }}
+  policyTypes:
+  - Ingress
+  - Egress
+  ingress: {}
```
Comment (Mansalu) [inline stable/nginx/nginx/templates/NetworkPolicy.yaml:47]:
```suggestion
  egress:
  - {}
  ingress:
  - {}
```

Need to allow ingress and egress

File: stable/nginx/nginx/templates/NetworkPolicy.yaml [hunk 9]
```diff
@@ -0,0 +1,55 @@
+{{ if .Values.NetworkPolicy.Enabled }}
+apiVersion: networking.k8s.io/v1
+kind: NetworkPolicy
+metadata:
+  name: {{ template "nginx-fullname" . }}
```
Comment (Mansalu) [inline stable/nginx/nginx/templates/NetworkPolicy.yaml:5]:
```suggestion
  name: {{ template "nginx.fullname" . }}
```

File: stable/nginx/nginx/templates/NetworkPolicy.yaml [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+  - from:
+    - ipBlock:
+        cidr: {{ .Values.NetworkPolicy.AllowedCIDR }}
+{{ end }}
+---
+{{ if not .Values.NetworkPolicy.Enabled }}
+apiVersion: networking.k8s.io/v1
+kind: NetworkPolicy
+metadata:
+  name: {{ template "nginx-fullname" . }}
```
Comment (Mansalu) [inline stable/nginx/nginx/templates/NetworkPolicy.yaml:35]:
```suggestion
  name: {{ template "nginx.fullname" . }}
```

**Comment by griffithscg** [comment]
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: default-deny-all
spec:
  podSelector: {}
  policyTypes:
  - Ingress
  - Egress

**Comment by cnweaver** [comment]
This basically works as advertised, but a few details felt a little off when I tried it:

In some sense, I shouldn't be surprised, since leaving NetworkPolicy.Enabled set to `false` is documented to give universal access, but I hadn't expected it to do so by explicitly creating a network policy which whitelists all connections. I spent some time being confused as to why my global default deny policy didn't seem to work, since my nginx deployed 'without' its own policy continued to be accessible right through it. My expectation was that not enabling this feature would simply not create any policy object. 

When I tried limiting incoming connections to only one IP address (my home network's external Comcast address), it mostly worked as expected, in that from home I could still reach nginx, while from another Slate VM at U. Chicago I could not. However, I was again surprised because the first thing I tried was making a request from the same VM on which the Kubernetes cluster was running, expecting it to be blocked, but it was not. I assume that this probably has to do with the request coming from a different interface, or in some other way bypassing whatever Calico does to filter traffic. This seems like a loophole which needs to be documented somewhere.

**Comment by Mansalu** [comment]
Hmm. The second point here surprises me. From my understanding, network policy is just as much designed to protect you from entities inside the cluster as outside. Perhaps there is more configuration needed. I'd like to test if an arbitrary pod somewhere else on the cluster can reach the nginx with policy in place.

As for your first point, I think we can rework the policy so that it behaves more desirably. Not deploying any policy whatsoever if policy is disabled seems reasonable to me.

**Pull request merged**